File: python/day7/day7.py
import re
import heapq
from time import time
from collections import deque, defaultdict, Counter


input = open("../../inputs/day7.txt").readlines()


def part1(lines):
    ans = 0
    for line in lines:
        target, nums = line.split(":")
        target = int(target)
        nums = [int(x) for x in nums.split()]
        idx = 1
        s = nums[0]
        q = deque([(s, idx)])
        while q:
            sub_s, c_idx = q.popleft()
            if c_idx >= len(nums):
                continue
            r = nums[c_idx]
            add = sub_s + r
            mul = sub_s * r
            if (c_idx == len(nums) - 1) and target in [add, mul]:
                ans += target
                break
            if add <= target:
                q.append((add, c_idx + 1))
            if mul <= target:
                q.append((mul, c_idx + 1))

    return ans


# 249943303573988 too high
# 249908125305812 too low
# 249943041417600 correct but I changed nothing....
def part2(lines):
    ans = 0
    for line in lines:
        target, nums = line.split(":")
        target = int(target)
        nums = [int(x) for x in nums.split()]
        idx = 1
        s = nums[0]
        q = deque([(s, idx)])
        while q:
            sub_s, c_idx = q.popleft()
            if c_idx >= len(nums):
                break
            r = nums[c_idx]
            add = sub_s + r
            mul = sub_s * r
            comb = int(str(sub_s) + str(r))
            if (c_idx == len(nums) - 1) and target in [add, mul, comb]:
                ans += target
                break
            if add <= target:
                q.append((add, c_idx + 1))
            if mul <= target:
                q.append((mul, c_idx + 1))
            if comb <= target:
                q.append((comb, c_idx + 1))
    return ans


s = time()
print(part1(input))
print(f"took: {time() - s}")
print("********************")
s = time()
print(part2(input))
print(f"took: {time() - s}")

File: rust/src/bin/day7.rs
use std::cmp;
use std::collections::HashMap;
use std::collections::HashSet;
use std::collections::VecDeque;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let lines: Vec<String> = read_lines("../inputs/day7.txt")
        .unwrap()
        .map_while(Result::ok)
        .collect();
    part1(&lines);
    part2(&lines);
}
fn is_in_bounds(xi: i64, yi: i64, max_x: i64, max_y: i64) -> bool {
    (0 <= xi && xi <= max_x) && (0 <= yi && yi <= max_y)
}

fn part1(lines: &[String]) {
    let mut ans = 0;
    lines.iter().for_each(|line| {
        let mut splits = line.split(":");
        let target: i64 = splits.next().unwrap().parse::<i64>().unwrap();
        let nums_s = splits.next().unwrap();
        let mut nums: Vec<i64> = Vec::new();
        nums_s.split_whitespace().for_each(|num| {
            nums.push(num.parse::<i64>().unwrap());
        });
        let idx = 1i64;
        let mut q: VecDeque<(i64, i64)> = VecDeque::new();
        q.push_back((nums[0], idx));
        while !q.is_empty() {
            let (sub_s, c_idx) = q.pop_front().unwrap();
            if c_idx >= nums.len() as i64 {
                continue;
            }
            let r = nums[c_idx as usize];
            let add = sub_s + r;
            let mul = sub_s * r;
            if (c_idx == (nums.len() - 1) as i64) && (target == add || target == mul) {
                ans += target;
                break;
            }
            if add <= target {
                q.push_back((add, c_idx + 1));
            }
            if mul <= target {
                q.push_back((mul, c_idx + 1));
            }
        }
    });
    println!("ans: {}", ans);
}
fn part2(lines: &[String]) {
    let mut ans = 0;
    lines.iter().for_each(|line| {
        let mut splits = line.split(":");
        let target: i64 = splits.next().unwrap().parse::<i64>().unwrap();
        let nums_s = splits.next().unwrap();
        let mut nums: Vec<i64> = Vec::new();
        nums_s.split_whitespace().for_each(|num| {
            nums.push(num.parse::<i64>().unwrap());
        });
        let idx = 1i64;
        let mut q: VecDeque<(i64, i64)> = VecDeque::new();
        q.push_back((nums[0], idx));
        while !q.is_empty() {
            let (sub_s, c_idx) = q.pop_front().unwrap();
            if c_idx >= nums.len() as i64 {
                continue;
            }
            let r = nums[c_idx as usize];
            let add = sub_s + r;
            let mul = sub_s * r;
            let comb: i64 = format!("{}{}", sub_s, r).parse().unwrap();
            if (c_idx == (nums.len() - 1) as i64)
                && (target == add || target == mul || target == comb)
            {
                ans += target;
                break;
            }
            if add <= target {
                q.push_back((add, c_idx + 1));
            }
            if mul <= target {
                q.push_back((mul, c_idx + 1));
            }
            if comb <= target {
                q.push_back((comb, c_idx + 1));
            }
        }
    });
    println!("ans: {}", ans);
}
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}
